Photon: Do not image_downsize for REST API requests in the edit context

kraftbj opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

Fixes #10580

#### Changes proposed in this Pull Request:
* Return out of image_downsize early for REST API edit context requests to media endpoints

#### Testing instructions:
* See issue above.

#### Proposed changelog entry for your changes:
* Improved compatibility for the image CDN with the new WordPress Block Editor.

kraftbj committed:
Do not image_downsize for REST API requests in the edit context

mdawaffe commented on class.photon.php:
```diff
@@ -473,6 +473,34 @@ public function filter_image_downsize( $image, $attachment_id, $size ) {
+				 *
+				 * @since 6.7.1
+				 *
+				 * @param bool false Allow Photon to run on the REST API media endpoint. Default to false.
```
This filter claims to be able to adjust photon settings during REST API requests, but it only fires if `?context=edit`.

mdawaffe commented on class.photon.php:
```diff
@@ -473,6 +473,34 @@ public function filter_image_downsize( $image, $attachment_id, $size ) {
+		// Similar to the need to skip the admin, we should skip the REST API to avoid Photonizing images out of context.
+		if ( defined( 'REST_REQUEST' ) && REST_REQUEST ) {
+			global $wp;
+			if ( false !== strpos( $wp->query_vars['rest_route'], '/wp/v2/media/' ) && ! empty( $_REQUEST['context'] ) && 'edit' === $_REQUEST['context'] && // phpcs:ignore WordPress.Security.NonceVerification.NoNonceVerification
```
Rather than all these conditionals, it may be simpler to attach something to the `rest_request_before_callbacks` filter (and cleanup via the `rest_request_after_callbacks` filter).

mdawaffe replied:
The `WP_REST_Request` object is passed to these filters, so we could more robustly check the `context` (which isn't always set via `$_GET` parameter: sometimes it's set programmatically, e.g., when calling `rest_do_request()`).

kraftbj replied:
Thank you! I thought there had to be a more elegant way of handling it and the callback filter was the missing link.

kraftbj replied:
Changed things up to use the `rest_request_before_callbacks`. I'm chewing on still how to clean it up by the way I did it.

mdawaffe reviewed (commented):
Some thoughts inline.

kraftbj committed:
Revert "Do not image_downsize for REST API requests in the edit context"

This reverts commit 083e91867a4e084372e18245f8a0f3c8d5bdfa97.

kraftbj committed:
PHPCS: Add brackets around conditional

kraftbj committed:
Photon: Do not fire on image_downsize for REST API requests on the wp/v2/media endpoint in the edit context

kraftbj commented:
Based on the docs, it should be an instance of the class checked as populated by the $endpoints property of the class. I need to hand this one off for testing and getting across the finish line. Thanks for your work on this!

mdawaffe committed:
Photon: Typo

mdawaffe committed:
Photon: Remove `$handler` check.

We're not checking for the right value.

Also, in theory, `$handler` could be any callable.

mdawaffe committed:
Photon: Clean up after ourselves after deciding not to Photonize during `?context=edit` REST requests.

Fix some PHPDoc.

mdawaffe committed:
Photon: Tests for URLs in REST API Responses

Clean the `Jetpack_Photon` state before each test.

The singleton/static class property make it annoying to clean the state.

jeherve commented on tests/php/test_class.jetpack_photon.php:
```diff
@@ -11,13 +11,30 @@ public function setUp() {
+		// static variable.
+		// l337 h4X0Ring required:
+		$instance = new ReflectionProperty( 'Jetpack_Photon', '__instance' );
+		$instance->setAccessible( true );
```
This makes our PHP 5.2 tests trip:

```
Fatal error: Call to undefined method ReflectionProperty::setAccessible() in wp-content/plugins/jetpack/tests/php/test_class.jetpack_photon.php on line 35
```

mdawaffe replied:
06e261d29 may "fix" it.

jeherve replied:
I am afraid the tests are still failing :( Travis won't let us merge that one.

mdawaffe replied:
OK - I'll look into it more later. Thanks for the ping!

mdawaffe replied:
Looks like it's working as of ea220aa376bc9eb2209aebce6193b50876b7122f.

jeherve reviewed (changes requested):
PHP 5.2 is causing us some issues here :(

mdawaffe committed:
Photon Tests: Require PHP 5.3

The singleton/static class property make it annoying to test, and is only
possible in PHP 5.3+

http://php.net/manual/en/reflectionproperty.setaccessible.php

mdawaffe committed:
Photon Tests: Fix Tests for PHP < 5.3 for real

The version of PHPUnit we're running on TravisCI is very old and doesn't
support `@requires`.

jeherve reviewed (approved):
It seems to work well in my tests. 🚢

jeherve commented:
Cherry-picked to `branch-6.8` in f2133f043495be31c0909f531d7d6d73eb1bfea0
